quinn/src/recv_stream.rs

```diff
@@ -284,6 +284,52 @@ impl RecvStream {
         self.stream
     }
 
+    /// Completes when the stream is reset by the peer or otherwise closed
+    ///
+    /// **Does not** actively cause the stream to become reset. Contrast [`stop()`](Self::stop),
+    /// which notifies the peer that the stream will no longer be read and should be reset.
+    ///
+    /// Yields `Some` with the reset error code when the stream is reset by the peer. Yields `None`
+    /// when the stream was previously [`stop()`](Self::stop)ed, or when the stream was
+    /// [`finish()`](crate::SendStream::finish)ed by the peer and all data has been received, after
+    /// which it is no longer meaningful for the stream to be reset.
+    ///
+    /// Notification of a reset is delivered exactly once. After a reset error code has been
+    /// delivered by *any* means, including calls to `reset()` or any operation that returns
+    /// [`ReadError::Reset`], future calls to `reset()` will return `Ok(None)`, and future calls to
+    /// methods that can return [`ReadError`] will return [`ReadError::ClosedStream`].
```

Comment:
I think it would be a bit confusing - might be better to explicitly maintain a notion that the error code has been handed over and return an error.
From the looks of it - it might not be that difficult to copy the error code out instead of consuming it.

---

Is this call cancellation-safe?

Reply:
Made resets persistent and documented cancel-safety.